li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -487,22 +494,52 @@ synchronized public void unsubscribe() {
 
   @Override
   synchronized public void commitSync() {
-    throw new UnsupportedOperationException("Not implemented yet");
+    commitSync(_defaultApiTimeout);
   }
 
   @Override
   synchronized public void commitSync(Duration timeout) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    executePerClusterCallback("commitSync", null /* call for all consumers */,
+        (consumer, ignored, remainingTimeout) -> consumer.commitSync(remainingTimeout), timeout);
   }
 
   @Override
   public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    commitSync(offsets, _defaultApiTimeout);
   }
 
+  // ATTN: UnknownTopicOrPartitionException may be received - this is a retriable exception..
+  // if not resolved by the time, timeout exception
   @Override
   synchronized public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets, Duration timeout) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    // Vanilla consumer's commitSync() retries on nonexistent partitions until the given timeout is reached. Since
+    // federated consumer does not know where the currently nonexistent topics in the input will be created, it mimics
+    // the vanilla consumer's behavior by retrying after waiting retry.backoff.ms until the given timeout is reached.
+    long now = System.currentTimeMillis();
+    long deadlineTimeMs = now + timeout.toMillis();
+    while (now < deadlineTimeMs) {
+      PartitionKeyedMapLookupResult offsetsByClusterResult = getPartitionKeyedMapsByCluster(offsets);
```

Comment:
if the while() loop executes multiple times (because some topics done exist?) it will commit the same offsets for existing topicPartitions N times. i think on successful commit the TPs committed should be taken out and not committed over again - its just wasteful